pkcs12/src/authenticated_safe.rs

```diff
@@ -0,0 +1,163 @@
+    pub fn content_type(&self) -> AuthenticatedSafeContentType {
+        match self {
+            Self::Data(_) => AuthenticatedSafeContentType::Data,
+            Self::EncryptedData(_) => AuthenticatedSafeContentType::EncryptedData,
+            Self::EnvelopedData(_) => AuthenticatedSafeContentType::EnvelopedData,
+        }
+    }
+}
+
+impl<'a> DecodeValue<'a> for AuthenticatedSafeItem<'a> {
+    fn decode_value<R: Reader<'a>>(
+        reader: &mut R,
+        header: Header,
+    ) -> der::Result<AuthenticatedSafeItem<'a>> {
+        reader.read_nested(header.length, |reader| {
+            let bag_type = reader.decode()?;
+            match bag_type {
+                AuthenticatedSafeContentType::Data => {
+                    let inner = reader
+                        .context_specific::<OctetStringRef<'a>>(CONTENT_TAG, TagMode::Explicit)?;
+                    let contents = SafeContents::from_der(inner.unwrap().as_bytes())?;
+                    Ok(AuthenticatedSafeItem::Data(Some(contents)))
+                }
+                AuthenticatedSafeContentType::EncryptedData => {
+                    Ok(AuthenticatedSafeItem::EncryptedData(
+                        reader.context_specific::<EncryptedDataContent<'a>>(
+                            CONTENT_TAG,
+                            TagMode::Explicit,
+                        )?,
+                    ))
+                }
+                AuthenticatedSafeContentType::EnvelopedData => {
+                    Ok(AuthenticatedSafeItem::EnvelopedData(
+                        reader
+                            .context_specific::<SafeContents<'a>>(CONTENT_TAG, TagMode::Explicit)?,
+                    ))
+                }
+            }
+        })
+    }
+}
+
+impl<'a> Sequence<'a> for AuthenticatedSafeItem<'a> {
+    fn fields<F, T>(&self, f: F) -> der::Result<T>
+    where
+        F: FnOnce(&[&dyn Encode]) -> der::Result<T>,
+    {
+        match self {
+            AuthenticatedSafeItem::Data(data) => f(&[
+                &self.content_type(),
+                &data.as_ref().map(|d| ContextSpecific {
+                    tag_number: CONTENT_TAG,
+                    tag_mode: TagMode::Explicit,
+                    value: d.clone(),
+                }),
+            ]),
+            AuthenticatedSafeItem::EncryptedData(data) => f(&[
+                &self.content_type(),
+                &data.as_ref().map(|d| ContextSpecific {
+                    tag_number: CONTENT_TAG,
+                    tag_mode: TagMode::Explicit,
+                    value: d.clone(),
+                }),
+            ]),
+            AuthenticatedSafeItem::EnvelopedData(data) => f(&[
+                &self.content_type(),
+                &data.as_ref().map(|d| ContextSpecific {
+                    tag_number: CONTENT_TAG,
+                    tag_mode: TagMode::Explicit,
+                    value: d.clone(),
+                }),
+            ]),
+        }
+    }
```

Comment:
FYI, this method was removed in #828 

Here's a similar example, although it could probably be cleaned up a bit / DRYed out: https://github.com/RustCrypto/formats/pull/828/files#diff-7eada31819a5738c81669b63773252d75c0ae287f02296637ee5862e36b18408